Add annotations to Painless whitelist

jdconrad opened the pull request:
This change adds the ability to attach annotative information for classes, methods, fields, static methods, class bindings, and instance bindings during Painless whitelisting.

Annotations are specified as ```@annotation``` or optionally as ```@annotation[some info]```.

Annotations open up the ability to specify whitelist objects as having a short name (```no_import -> @no_import```) or deprecated.

jdconrad committed:
Start to parse annotations in the whitelist

jdconrad committed:
Finish parsing annotations

jdconrad committed:
Finish annotation parsing

jdconrad committed:
Add docs

elasticmachine commented:
Pinging @elastic/es-core-infra

jdconrad committed:
Merge branch 'master' into annotate

rjernst commented on modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistInstanceBinding.java:
```diff
@@ -47,15 +49,21 @@
     public final List<String> canonicalTypeNameParameters;
 
+    /** The {@link Map} of annotations for this field. */
+    public final Map<String, String> annotations;
```
I would do this a bit differently. A couple suggestions:

* have a WhitelistAnnotation marker interface
* have annotation parsers keyed by the string we look for in the whitelist, eg deprecation
* on the whitelist objects, we can have a Map<Class<? extends Annotation>, ? extends Annotation>, so that we can look up annnotations by type
* pass the map of parsers to the whitelist loader, and have a static base annotations that can be used, eg with `no_import` and in the future `deprecated`
* pass the annotation parser a Map<String, String>, ie make the contents of the parsed annotation a keyed map like java has. this will allow future expansion and clarity with named parameters instead of positional

jdconrad replied:
Thanks for the feedback.  Will get to work with the changes we discussed.

rjernst reviewed (commented):
Just one general suggestion about some better flexibility. I think what I suggest will allow us to do better validation at whitelist parsing time of values passed into an annotation.

jdconrad committed:
Merge branch 'master' into annotate

jdconrad committed:
Merge branch 'master' into annotate

jdconrad committed:
Parser changes.

jdconrad committed:
Merge branch 'master' into annotate

jdconrad committed:
Add annotation parser.

jdconrad committed:
Merge branch 'master' into annotate

jdconrad committed:
Add annotation tests for Painless core.

jdconrad committed:
Merge branch 'master' into annotate

jdconrad committed:
Merge branch 'master' into annotate

jdconrad committed:
Add example annotation to Painless example extension.

jdconrad commented:
@rjernst I think I've addressed all the changes talked about, so this is ready for the next round of review.  I also added a large number of tests, and added an example annotation to the example Painless extension.

rjernst commented on modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/annotation/NoImportAnnotationParser.java:
```diff
@@ -0,0 +1,40 @@
+            throw new IllegalArgumentException("unexpected parameters for [@no_import] annotation, found " + arguments);
+        }
+
+        return new NoImportAnnotation();
```
This could be a static instance?

jdconrad replied:
Changed.

rjernst commented on modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistField.java:
```diff
@@ -37,10 +41,17 @@
     /** Standard constructor.  All values must be not {@code null}. */
-    public WhitelistField(String origin, String fieldName, String canonicalTypeNameParameter) {
+    public WhitelistField(String origin, String fieldName, String canonicalTypeNameParameter,
+            Map<Class<?>, PainlessAnnotation> painlessAnnotations) {
```
I think we should pass in a list of annotations, and build the map ourselves internally. This way a user manually constructing a whitelist (not using the parser) can't mess up the map.

jdconrad replied:
Done.

rjernst commented on modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/annotation/PainlessAnnotation.java:
```diff
@@ -0,0 +1,25 @@
+
+package org.elasticsearch.painless.spi.annotation;
+
+public interface PainlessAnnotation {
```
I'm not sure we need a base class. There are 2 cases we might want to use a name. The first is already the key of the map we pass into the whitelist parser. The second is if we want to add these annotations to the documentation of the whitelist. There I think we can use toString()? Even if we do eventually need a base class, let's add that when we have a concrete use for it.

jdconrad replied:
Done.

rjernst reviewed (approved):
LGTM, just a couple more suggestions

jdconrad committed:
Response to PR comments.

jdconrad commented:
@rjernst Thanks for the reviews!  Will commit as soon as CI passes.

jdconrad commented:
@elasticmachine run elasticsearch-ci/1

jdconrad committed:
Merge branch 'master' into annotate
